File: backend/app/api/attribution.py
"""Performance attribution API (GET /api/attribution/*).

Read-only decomposition of realized PnL across symbols, directions, exit
reasons, market sessions and calendar days. Sourced entirely from the
``OrderRecord`` exit ledger; never writes.
"""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.api.auth import require_api_key
from app.database import get_db
from app.services.performance_attribution_service import PerformanceAttributionService

router = APIRouter(
    prefix="/api/attribution",
    tags=["attribution"],
    dependencies=[Depends(require_api_key())],
)


# ----------------------------------------------------------------------
# response schemas (local to this router per AGENTS.md)
# ----------------------------------------------------------------------


class SymbolBucket(BaseModel):
    total_pnl: float
    trade_count: int
    win_count: int
    avg_pnl: float


class DirectionBucket(BaseModel):
    total_pnl: float
    trade_count: int


class ExitReasonBucket(BaseModel):
    total_pnl: float
    trade_count: int


class SessionBucket(BaseModel):
    total_pnl: float
    trade_count: int


class DayBucket(BaseModel):
    date: str
    pnl: float
    trade_count: int


class AttributionResponse(BaseModel):
    period_days: int = Field(..., ge=0)
    total_pnl: float
    total_trades: int = Field(..., ge=0)
    win_rate: float = Field(..., ge=0.0, le=1.0)
    by_symbol: dict[str, SymbolBucket]
    by_direction: dict[str, DirectionBucket]
    by_exit_reason: dict[str, ExitReasonBucket]
    by_session: dict[str, SessionBucket]
    by_day: list[DayBucket]


class TopContributorRow(BaseModel):
    symbol: str
    total_pnl: float
    trade_count: int = Field(..., ge=0)
    win_rate: float = Field(..., ge=0.0, le=1.0)
    avg_holding_minutes: float = Field(..., ge=0.0)


# ----------------------------------------------------------------------
# endpoints (sync handlers per project convention)
# ----------------------------------------------------------------------


@router.get("/pnl", response_model=AttributionResponse)
def get_pnl_attribution(
    days: int = Query(default=30, ge=1, le=3650, description="Lookback window in days"),
    symbol: str | None = Query(
        default=None,
        description="Restrict to a single symbol (exact match, e.g. AAPL.US)",
    ),
    db: Session = Depends(get_db),
) -> dict[str, Any]:
    """Multi-dimensional realized-PnL breakdown for the lookback window."""
    return PerformanceAttributionService(db).attribute_pnl(days=days, symbol=symbol)


@router.get("/top-contributors", response_model=list[TopContributorRow])
def get_top_contributors(
    days: int = Query(default=30, ge=1, le=3650, description="Lookback window in days"),
    limit: int = Query(default=10, ge=1, le=200, description="Maximum rows to return"),
    db: Session = Depends(get_db),
) -> list[dict[str, Any]]:
    """Top symbols by absolute PnL contribution, with holding-time stats."""
    return PerformanceAttributionService(db).top_contributors(days=days, limit=limit)

File: frontend/src/api/attribution.ts
import { api } from './client'

export interface PnlBucket {
  total_pnl: number
  trade_count: number
  win_count?: number
  avg_pnl?: number
}

export interface DayPnl {
  date: string
  pnl: number
  trade_count: number
}

export interface AttributionResult {
  period_days: number
  total_pnl: number
  total_trades: number
  win_rate: number
  by_symbol: Record<string, PnlBucket>
  by_direction: Record<string, PnlBucket>
  by_exit_reason: Record<string, PnlBucket>
  by_session: Record<string, PnlBucket>
  by_day: DayPnl[]
}

export interface TopContributor {
  symbol: string
  total_pnl: number
  trade_count: number
  win_rate: number
  avg_holding_minutes: number
}

export async function getPnlAttribution(params: { days?: number; symbol?: string } = {}): Promise<AttributionResult> {
  const resp = await api.get('/api/attribution/pnl', { params })
  return resp.data
}

export async function getTopContributors(params: { days?: number; limit?: number } = {}): Promise<TopContributor[]> {
  const resp = await api.get('/api/attribution/top-contributors', { params })
  return resp.data
}
